src/creature_instances.c

```diff
@@ -537,6 +548,133 @@ long process_creature_self_spell_casting(struct Thing* creatng)
+    {
+        return CrInst_NULL;
+    }
+
+    if (!creature_instance_is_available(creatng, CrInst_RANGED_HEAL) ||
+        !creature_instance_has_reset(creatng, CrInst_RANGED_HEAL))
+    {
+        return CrInst_NULL;
+    }
+
+    struct Thing* best_choice = NULL;
+    if (creature_would_benefit_from_healing(creatng))
+    {
+        // Save itself first.
+        best_choice = creatng;
+    }
+    else
+    {
+        // Search nearby area, collect candidates.
+        struct Dungeon* dungeon = get_players_num_dungeon(creatng->owner);
+        int creature_idx = dungeon->creatr_list_start;
+        int k = 0;
+        while (creature_idx != 0)
+        {
+            struct Thing* candidate_thing = thing_get(creature_idx);
+            struct CreatureControl* candidate_ctrl = creature_control_get_from_thing(candidate_thing);
+            creature_idx = candidate_ctrl->players_next_creature_idx;
+            if (creature_control_invalid(candidate_ctrl))
+            {
+                ERRORLOG("Invalid creature control");
+                break;
+            }
+            // Check validation of candidate's state.
+            if (candidate_thing->index == creatng->index ||
+                creature_is_dying(candidate_thing) ||
+                thing_is_picked_up(candidate_thing) || creature_is_being_dropped(candidate_thing) ||
+                creature_is_being_sacrificed(candidate_thing) || creature_is_being_summoned(candidate_thing) ||
+                // Creature who is leaving doesn't deserve heal from allies.
+                creatng->continue_state == CrSt_CreatureLeaves ||
+                creatng->active_state == CrSt_CreatureLeavingDungeon ||
+                creatng->active_state == CrSt_CreatureScavengedDisappear ||
+                // Candidate shouldn't be fight with the caster.
+                creature_has_creature_in_combat(creatng, candidate_thing) ||
+                creature_has_creature_in_combat(candidate_thing, creatng) ||
+                !creature_would_benefit_from_healing(candidate_thing))
+            {
+                continue;
+            }
+
+            // We don't check the field of view of caster here because the following reasons.
+            // a. The wounded creature can speak to get attention from the caster.
+            // b. The thing_in_field_of_view() is bugged.
+            // c. The caster will turn to face the target, so it is visually fine.
+            int range = get_combat_distance(creatng, candidate_thing);
+            const struct InstanceInfo* inst_inf = creature_instance_info_get(CrInst_RANGED_HEAL);
+            // Check validation of candidate's spatial status.
+            if (range >= inst_inf->range_min && range <= inst_inf->range_max &&
+                creature_can_see_combat_path(creatng, candidate_thing, range))
+            {
+                if (best_choice == NULL || candidate_thing->health < best_choice->health)
```

Comment:
I think the health percentage would be better over the absolute health. Also, brackets would make this clearer.

Reply:
OK.